网关使用spring-cloud-starter-gateway-mvc整合polaris问题

spring cloud 整合北极星版本1.13.2-2023.0.0
网关使用spring-cloud-starter-gateway-mvc 异常。

完整异常信息如下：

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
启动ApplicationContext时出错。若要显示条件评估报告，请在启用“debug”的情况下重新运行应用程序。

2024-07-19T10:16:49.665+08:00 ERROR 64964 --- [aurofa-commons] [ main] o.s.boot.SpringApplication : Application run failed
2012-07-19 t10:16:49 .665 08:00 ERROR 64964—[aurora -commons] [main] o.s.boot.SpringApplication:应用程序运行失败

java.lang.IllegalStateException: Error processing condition on com.tencent.cloud.polaris.DiscoveryPropertiesAutoConfiguration.polarisDiscoveryProperties
java.lang.IllegalStateException: com.tencent.cloud.polaris.DiscoveryPropertiesAutoConfiguration.polarisDiscoveryProperties的错误处理条件

at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:60) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
[spring-boot- autoconfiguration -3.3.1.jar:3.3.1]

at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:108) ~[spring-context-6.1.10.jar:6.1.10]
[springframework.context.annotation. conditionevaluator . shouldskip (ConditionEvaluator.java:108) ~[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForBeanMethod(ConfigurationClassBeanDefinitionReader.java:183) ~[spring-context-6.1.10.jar:6.1.10]
[springframework.context.annotation. configurationclassbeandefinitionreader . loadbeandefinitionsforbeanmethod] [spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:144) ~[spring-context-6.1.10.jar:6.1.10]
[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:120) ~[spring-context-6.1.10.jar:6.1.10]
[springframework.context.annotation. ConfigurationClassBeanDefinitionReader.java:120]

at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:429) ~[spring-context-6.1.10.jar:6.1.10]
[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:290) ~[spring-context-6.1.10.jar:6.1.10]
在org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:290) ~[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:349) ~[spring-context-6.1.10.jar:6.1.10]
PostProcessorRegistrationDelegate.java:349 ~[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:118) ~[spring-context-6.1.10.jar:6.1.10]
PostProcessorRegistrationDelegate.java:118 ~[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:788) ~[spring-context-6.1.10.jar:6.1.10]
AbstractApplicationContext.java:788) ~[spring-context-6.1.10.jar:6.1.10]

at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:606) ~[spring-context-6.1.10.jar:6.1.10]
AbstractApplicationContext.java:606 ~[spring-context-6.1.10.jar:6.1.10]

at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.3.1.jar:3.3.1]
[spring-boot-3.3.1.jar:3.3.1]

at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754) ~[spring-boot-3.3.1.jar:3.3.1]
[spring-boot-3.3.1.jar:3.3.1]

at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456) ~[spring-boot-3.3.1.jar:3.3.1]
在org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456) ~[springboot -3.3.1.jar:3.3.1]

at org.springframework.boot.SpringApplication.run(SpringApplication.java:335) ~[spring-boot-3.3.1.jar:3.3.1]
[spring-boot-3.3.1.jar:3.3.1]

at org.springframework.boot.SpringApplication.run(SpringApplication.java:1363) ~[spring-boot-3.3.1.jar:3.3.1]
[spring-boot-3.3.1.jar:3.3.1]

at org.springframework.boot.SpringApplication.run(SpringApplication.java:1352) ~[spring-boot-3.3.1.jar:3.3.1]
[spring-boot-3.3.1.jar:3.3.1]

at com.aurora.framework.gateway.AuroraGatewayApplication.main(AuroraGatewayApplication.java:11) ~[classes/:na]
[classes/:na]

Caused by: java.lang.IllegalStateException: Failed to introspect Class [com.tencent.cloud.plugin.gateway.SCGPluginsAutoConfiguration$RuleStainingPluginConfiguration] from ClassLoader [jdk.internal.loader.ClassLoaders$AppClassLoader@4e0e2f2a]
原因:java.lang.IllegalStateException: Failed to introspecintrospection Class [com.tencent.cloud.plugin.gateway.]SCGPluginsAutoConfiguration$RuleStainingPluginConfiguration] from ClassLoader [jdk.internal.loader.ClassLoader $AppClassLoader@4e0e2f2a]

at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:483) ~[spring-core-6.1.10.jar:6.1.10]
在org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:483) ~[spring-core-6.1.10.jar:6.1.10]

at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:360) ~[spring-core-6.1.10.jar:6.1.10]
在org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:360) ~[spring-core-6.1.10.jar:6.1.10]

at org.springframework.util.ReflectionUtils.getUniqueDeclaredMethods(ReflectionUtils.java:417) ~[spring-core-6.1.10.jar:6.1.10]
在org.springframework.util. reflectionutils . getuniquedeclardmethods (ReflectionUtils.java:417) ~[spring-core-6.1.10.jar:6.1.10]

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.lambda$getTypeForFactoryMethod$1(AbstractAutowireCapableBeanFactory.java:750) ~[spring-beans-6.1.10.jar:6.1.10]
$ getypeforfactorymethod $1(AbstractAutowireCapableBeanFactory.java:750) ~[spring-beans-6.1.10.jar:6.1.10]

at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708) ~[na:na]
和 java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent（ConcurrentHashMap.java：1708） ~[that：na]

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getTypeForFactoryMethod(AbstractAutowireCapableBeanFactory.java:749) ~[spring-beans-6.1.10.jar:6.1.10]
@ org.springframework.beans.factory. abstractautowirecapablebeanfactory . gettypeforfactorymethod (AbstractAutowireCapableBeanFactory.java:749) ~

at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:682) ~[spring-beans-6.1.10.jar:6.1.10]
at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:653) ~[spring-beans-6.1.10.jar:6.1.10]
at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1687) ~[spring-beans-6.1.10.jar:6.1.10]
at org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:562) ~[spring-beans-6.1.10.jar:6.1.10]
at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:534) ~[spring-beans-6.1.10.jar:6.1.10]
at org.springframework.boot.autoconfigure.condition.OnBeanCondition.collectBeanNamesForType(OnBeanCondition.java:247) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getBeanNamesForType(OnBeanCondition.java:240) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getBeanNamesForType(OnBeanCondition.java:230) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchingBeans(OnBeanCondition.java:183) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchOutcome(OnBeanCondition.java:158) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:47) ~[spring-boot-autoconfigure-3.3.1.jar:3.3.1]
... 17 common frames omitted
Caused by: java.lang.NoClassDefFoundError: org/springframework/cloud/gateway/filter/GlobalFilter
at java.base/java.lang.ClassLoader.defineClass1(Native Method) ~[na:na]
at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027) ~[na:na]
at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150) ~[na:na]
at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862) ~[na:na]
at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760) ~[na:na]
at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681) ~[na:na]
at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639) ~[na:na]
at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[na:na]
at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526) ~[na:na]
at java.base/java.lang.Class.getDeclaredMethods0(Native Method) ~[na:na]
at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3578) ~[na:na]
at java.base/java.lang.Class.getDeclaredMethods(Class.java:2676) ~[na:na]
at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:465) ~[spring-core-6.1.10.jar:6.1.10]
... 33 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.springframework.cloud.gateway.filter.GlobalFilter
at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[na:na]
at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[na:na]
at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526) ~[na:na]
... 46 common frames omitted
…46个常用框架省略

这是polaris和mvc模式的gateway还没集成好吧？

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,3 +16,4 @@
 - [fix:fix no registry when lossless is disabled.](https://github.com/Tencent/spring-cloud-tencent/pull/1312)
 - [fix:fix the ratelimit bug for 2023](https://github.com/Tencent/spring-cloud-tencent/pull/1316)
 - [fix:update spring boot version](https://github.com/Tencent/spring-cloud-tencent/pull/1333)
+- [feat:add scg-mvc example](https://github.com/Tencent/spring-cloud-tencent/pull/1367)

--- a/spring-cloud-tencent-examples/quickstart-example/pom.xml
+++ b/spring-cloud-tencent-examples/quickstart-example/pom.xml
@@ -19,5 +19,6 @@
 		<module>quickstart-caller-service</module>
 		<module>quickstart-callee-service-a</module>
 		<module>quickstart-callee-service-b</module>
+		<module>quickstart-gateway-mvc-service</module>
 	</modules>
 </project>

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/pom.xml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/pom.xml
@@ -11,7 +11,7 @@
 	<modelVersion>4.0.0</modelVersion>
 
 	<artifactId>quickstart-gateway-mvc-service</artifactId>
-	<name>Quickstart Gateway Service</name>
+	<name>Quickstart Gateway MVC Service</name>
 
 	<dependencies>
 		<dependency>

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/pom.xml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/pom.xml
@@ -1,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		 xmlns="http://maven.apache.org/POM/4.0.0"
+		 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<parent>
+		<artifactId>quickstart-example</artifactId>
+		<groupId>com.tencent.cloud</groupId>
+		<version>${revision}</version>
+		<relativePath>../pom.xml</relativePath>
+	</parent>
+	<modelVersion>4.0.0</modelVersion>
+
+	<artifactId>quickstart-gateway-mvc-service</artifactId>
+	<name>Quickstart Gateway Service</name>
+
+	<dependencies>
+		<dependency>
+			<groupId>com.tencent.cloud</groupId>
+			<artifactId>spring-cloud-starter-tencent-all</artifactId>
+		</dependency>
+
+		<dependency>
+			<groupId>com.tencent.cloud</groupId>
+			<artifactId>spring-cloud-tencent-gateway-plugin</artifactId>
+		</dependency>
+
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-starter-gateway-mvc</artifactId>
+		</dependency>
+
+	</dependencies>
+
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.springframework.boot</groupId>
+				<artifactId>spring-boot-maven-plugin</artifactId>
+				<executions>
+					<execution>
+						<goals>
+							<goal>repackage</goal>
+						</goals>
+					</execution>
+				</executions>
+			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-source-plugin</artifactId>
+				<version>3.2.0</version>
+				<executions>
+					<execution>
+						<id>attach-sources</id>
+						<goals>
+							<goal>jar</goal>
+						</goals>
+					</execution>
+				</executions>
+			</plugin>
+		</plugins>
+	</build>
+</project>

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/pom.xml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/pom.xml
@@ -20,11 +20,6 @@
 		</dependency>
 
 		<dependency>
-			<groupId>com.tencent.cloud</groupId>
-			<artifactId>spring-cloud-tencent-gateway-plugin</artifactId>
-		</dependency>
-
-		<dependency>
 			<groupId>org.springframework.cloud</groupId>
 			<artifactId>spring-cloud-starter-gateway-mvc</artifactId>
 		</dependency>

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/java/com/tencent/cloud/quickstart/gatewaymvc/FallbackController.java
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/java/com/tencent/cloud/quickstart/gatewaymvc/FallbackController.java
@@ -1,0 +1,37 @@
+/*
+ * Tencent is pleased to support the open source community by making Spring Cloud Tencent available.
+ *
+ * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
+ *
+ * Licensed under the BSD 3-Clause License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.tencent.cloud.quickstart.gatewaymvc;
+
+import reactor.core.publisher.Mono;
+
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * FallbackController.
+ *
+ * sean yu
+ */
+@RestController
+public class FallbackController {
+
+	@GetMapping("/polaris-fallback")
+	Mono<String> getFallback() {
+		return Mono.just("fallback: trigger the refuse for service caller.");
+	}
+}

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/java/com/tencent/cloud/quickstart/gatewaymvc/QuickstartGatewayMvcApplication.java
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/java/com/tencent/cloud/quickstart/gatewaymvc/QuickstartGatewayMvcApplication.java
@@ -1,0 +1,35 @@
+/*
+ * Tencent is pleased to support the open source community by making Spring Cloud Tencent available.
+ *
+ * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
+ *
+ * Licensed under the BSD 3-Clause License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.tencent.cloud.quickstart.gatewaymvc;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+/**
+ * Quickstart SCG application.
+ *
+ * @author Haotian Zhang
+ */
+@SpringBootApplication
+public class QuickstartGatewayMvcApplication {
+
+	public static void main(String[] args) {
+		SpringApplication.run(QuickstartGatewayMvcApplication.class, args);
+	}
+
+}

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
@@ -1,0 +1,35 @@
+server:
+  port: 48080
+spring:
+  application:
+    name: QuickStartGatewayService
+  config:
+    import: optional:polaris
+  cloud:
+    tencent:
+      plugin:
+        router:
+          feature-env:
+            enabled: true
+    polaris:
+      address: grpc://119.91.66.223:8091
+      namespace: default
+      enabled: true
+      contract:
+        exposure: true
+        report:
+          enabled: true
+      stat:
+        enabled: true
+        port: 28081
+    gateway:
+      mvc:
+        routes:
+          - id: QuickstartCallerService
+            uri: lb://QuickstartCallerService
+            predicates:
+              - Path=/QuickstartCallerService/**
+            filters:
+              - StripPrefix=1
+
+

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
@@ -16,7 +16,7 @@ spring:
           enabled: true
       stat:
         enabled: true
-        port: 28081
+        port: 28085
     gateway:
       mvc:
         routes:

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
@@ -1,16 +1,11 @@
 server:
-  port: 48080
+  port: 48085
 spring:
   application:
     name: QuickStartGatewayMvcService
   config:
     import: optional:polaris
   cloud:
-    tencent:
-      plugin:
-        router:
-          feature-env:
-            enabled: true
     polaris:
       address: grpc://119.91.66.223:8091
       namespace: default

--- a/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
+++ b/spring-cloud-tencent-examples/quickstart-example/quickstart-gateway-mvc-service/src/main/resources/application.yml
@@ -2,7 +2,7 @@ server:
   port: 48080
 spring:
   application:
-    name: QuickStartGatewayService
+    name: QuickStartGatewayMvcService
   config:
     import: optional:polaris
   cloud:

